pkg/agent/cache/manager.go

```diff
@@ -0,0 +1,372 @@
+			m.Shutdown(err)
+			return
+		}
+		svid := resp.SvidUpdate.Svids[req.entry.RegistrationEntry.SpiffeId]
+		cert, err := x509.ParseCertificate(svid.SvidCert)
+		if err != nil {
+			stream.CloseSend()
+			m.Shutdown(err)
+			return
+		}
+		m.log.Debug("Sending to regEntries Channel: ", "entries: ", resp.SvidUpdate.RegistrationEntries)
+
+		m.regEntriesCh <- resp.SvidUpdate.RegistrationEntries
+		req.entry.SVID = svid
+		req.entry.Expiry = cert.NotAfter
+		m.log.Debug("Sending to CacheEntry Channel: ", "req: ", req)
+		m.cacheEntryCh <- req.entry
+
+	}
+	if err := stream.CloseSend(); err != nil {
+		m.Shutdown(err)
+		return
+	}
+	m.log.Debug("Done with fetchSVID")
+}
+
+func (m *manager) fetchWithEmptyCSR(svid []byte, key *ecdsa.PrivateKey) {
+	conn, err := m.getGRPCConn(svid, key)
+	if err != nil {
+		m.Shutdown(err)
+		return
+	}
+	stream, err := node.NewNodeClient(conn).FetchSVID(context.Background())
+	if err != nil {
+		m.Shutdown(err)
+		return
+	}
+	err = stream.Send(&node.FetchSVIDRequest{})
+	if err != nil {
+		m.Shutdown(err)
+		return
+	}
+	strmch := make(chan struct{})
+	go func() {
+		for {
+			resp, err := stream.Recv()
+			//		m.log.Debug("regEntries:", resp.SvidUpdate.RegistrationEntries)
+			if err == io.EOF {
+				close(strmch)
+				m.log.Debug("closing stream")
+				return
+			}
+
+			if err != nil {
+				m.Shutdown(err)
+				close(strmch)
+				return
+			}
+
+			m.regEntriesCh <- resp.SvidUpdate.RegistrationEntries
+		}
+	}()
+	stream.CloseSend()
+	<-strmch
+}
+
+func (m *manager) expiredCacheEntryHandler(cacheFrequency time.Duration, wg *sync.WaitGroup) {
+	defer wg.Done()
+	ticker := time.NewTicker(cacheFrequency)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			entryRequestMap := make(map[string][]EntryRequest)
+			for _, entries := range m.managedCache.Entries() {
+				for _, entry := range entries {
+					if entry.Expiry.Sub(time.Now()) < time.Until(entry.Expiry)/2 {
+						privateKey, err := ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+						if err != nil {
+							m.Shutdown(err)
+							break
+						}
+						csr, err := util.MakeCSR(privateKey, entry.RegistrationEntry.SpiffeId)
+						entry.PrivateKey = privateKey
+						parentID := entry.RegistrationEntry.ParentId
+						entryRequestMap[parentID] = append(entryRequestMap[parentID], EntryRequest{csr, entry})
+					}
+					spiffeID := entry.RegistrationEntry.SpiffeId
+					m.spiffeIdEntryMap[spiffeID] = entry
+				}
+			}
+			if len(entryRequestMap) != 0 {
+				m.log.Debug("Fetching Expired Entries", "len(entryRequestMap):", len(entryRequestMap), "entryMap", entryRequestMap)
+				m.entryRequestCh <- entryRequestMap
+			}
+			vanityRecord := m.managedCache.Entry([]*proto.Selector{&proto.Selector{Type: "spiffe_id", Value: m.baseSPIFFEID}})
+
+			if vanityRecord != nil {
+				m.log.Debug("Fetching new reg Entries for vanity spiffeid:", vanityRecord[0].RegistrationEntry.SpiffeId)
+				m.log.Debug("len(entryRequestMap):", len(entryRequestMap))
```

Comment:
Maybe we can remove some of these debug lines? I assume they were there for... debugging :) but some of them will be hard to grok for folks who are not intimately familiar with this bit of the code. Also feels like the volume of debug statements from cache manager will drown out lines from other areas in the agent